### PR TITLE
Fix Head References

### PR DESCRIPTION
**Pull request by nisdas**
Previously we were referring to the head root and head slot via a field in the blockchain service and its corresponding root in `canonical_roots`. However this had a flaw, as it assumed that each slot had only 1 root. However during periods without finality and a lot of forks, each slot can have multiple blocks. 

This PR instead removes head slot and replaces with a head identifier, so that we can easily glean the current head slot and root from it. This also makes sure that we return the right data when forks occur.

### CODE REVIEW

File: beacon-chain/blockchain/chain_info_test.go [hunk 1]
```diff
@@ -127,22 +129,29 @@ func TestPrevJustifiedCheckpt_GenesisRootOk(t *testing.T) {
...
 }
 
 func TestHeadRoot_CanRetrieve(t *testing.T) {
-	c := &Service{canonicalRoots: make(map[uint64][]byte)}
-	c.headSlot = 100
-	c.canonicalRoots[c.headSlot] = []byte{'A'}
+	c := &Service{canonicalRoots: make(map[[40]byte]bool)}
+	identifier := [40]byte{}
+	root := [32]byte{'A'}
+	copy(identifier[:8], bytesutil.Bytes8(100))
```
Comment (shayzluf) [inline beacon-chain/blockchain/chain_info_test.go]:
```suggestion
        slot := 100
	copy(identifier[:8], bytesutil.Bytes8(slot))
```

File: beacon-chain/blockchain/chain_info.go [hunk 2]
```diff
@@ -5,6 +5,8 @@ import (
 	"context"
 	"time"
 
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+
```
Comment (terencechain) [inline beacon-chain/blockchain/chain_info.go]: goimports

File: beacon-chain/blockchain/head_test.go [hunk 3]
```diff
@@ -18,15 +20,19 @@ func TestSaveHead_Same(t *testing.T) {
 	defer testDB.TeardownDB(t, db)
 	service := setupBeaconChain(t, db)
 
-	service.headSlot = 0
+	identifier := [40]byte{}
 	r := [32]byte{'A'}
-	service.canonicalRoots[0] = r[:]
+
+	copy(identifier[:8], bytesutil.Bytes8(0))
```
Comment (terencechain) [inline beacon-chain/blockchain/head_test.go]: I think a lot of these can be refactored into its own function to reduce disparity down the line

**Comment by prestonvanloon** [comment]
What is the status of this? Was this resolved in #4869 ?

**Comment by nisdas** [comment]
Yes, that supersedes this PR. I am closing this

**Pull request closed without merging**